Development/development

Awesomeluke1994 opened the pull request.

Awesomeluke1994 committed:
init commit

Awesomeluke1994 committed:
init commit

Awesomeluke1994 committed:
init commit

Awesomeluke1994 committed:
init commit

Awesomeluke1994 committed:
merge conflict

Awesomeluke1994 committed:
Fixed some typings in the user controller and auth service
Added exclusion of idea from git ignore.

JamesMahy commented on prisma/migrations/20220726190348_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+-- CreateTable
+CREATE TABLE "Post" (
+    "id" SERIAL NOT NULL,
+    "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Would usually prefer DATETIME over TIMESTAMP, it makes data more readily accessible and use a trigger

eg

`createdAt` DATETIME NULL DEFAULT CURRENT_TIMESTAMP,
`updatedAt` DATETIME NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

Would also prefer `modifiedAt` over `updatedAt`

Awesomeluke1994 replied:
Ah these were the default migrations added prisma (removed them in further migrations) - DateTime or the lovely DateTimeOffset

JamesMahy commented on prisma/migrations/20220726190348_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+-- CreateTable
+CREATE TABLE "Post" (
```
would generally prefer lower case table names

JamesMahy commented on prisma/migrations/20220726190348_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+    "published" BOOLEAN NOT NULL DEFAULT false,
+    "authorId" INTEGER NOT NULL,
+
+    CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
```
would prefer lowercase constraint names

JamesMahy commented on prisma/migrations/20220726190348_init/migration.sql:
```diff
@@ -0,0 +1,42 @@
+);
+
+-- CreateIndex
+CREATE UNIQUE INDEX "Profile_userId_key" ON "Profile"("userId");
```
Would prefer lower case index names

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+import {CreateUserRequest, LoginRequest, TypedRequestBody} from "../global";
+import {Response} from "express";
```
Would prefer spacing after `{` and before `}` to aid readability

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+import {CreateUserRequest, LoginRequest, TypedRequestBody} from "../global";
```
you shouldn't need to import a global ts file for types, it should work automatically

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+import {Response} from "express";
+import {AuthService} from "../services/auth-service";
+
+const authService = new AuthService()
```
AuthService doesn't look like it retains a state, should it be exporting functions instead? There's no benefit to a "new" object

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+    registerUser = async (req: TypedRequestBody<CreateUserRequest>, res: Response) => {
+        const request = req.body
+
+        await authService.registerUser(request);
```
Missing error handling, what happens if this process fails?

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+export class AuthController {
+
+    registerUser = async (req: TypedRequestBody<CreateUserRequest>, res: Response) => {
+        const request = req.body
```
We should be validating the payload here, ensuring it conforms to the correct shape and no data is missing

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+    login = async(req: TypedRequestBody<LoginRequest>, res: Response) => {
+        const isASuccessfulLogin = await authService.login(req.body);
+
+        if(isASuccessfulLogin) {
```
spacing either side of `(` and `)`

JamesMahy commented on src/controllers/auth-controller.ts:
```diff
@@ -0,0 +1,27 @@
+    };
+
+    login = async(req: TypedRequestBody<LoginRequest>, res: Response) => {
+        const isASuccessfulLogin = await authService.login(req.body);
```
Needs error handling, we should also be validating the body conforms and only passing the relevant parameters to our service. 

Working this way too tightly couples the request to the service

JamesMahy commented on src/controllers/user-controller.ts:
```diff
@@ -0,0 +1,12 @@
+import {PrismaClient} from "@prisma/client";
```
spacing

JamesMahy commented on src/global.d.ts:
```diff
@@ -0,0 +1,17 @@
+import {Request} from "express";
```
Spacing

JamesMahy commented on src/controllers/user-controller.ts:
```diff
@@ -0,0 +1,12 @@
+const prisma = new PrismaClient()
+
+const getAll = async (req: any, res: any) => {
+    const allUsers = await prisma.user.findMany();
```
Feel like this endpoint should be protected?

JamesMahy commented on src/middleware/auth-middleware.ts:
```diff
@@ -0,0 +1,28 @@
+const jwt = require("jsonwebtoken")
```
commonjs?

JamesMahy commented on src/middleware/auth-middleware.ts:
```diff
@@ -0,0 +1,28 @@
+    } else {
+        res.status(403).send("Unauthorised");
+    }
+
```
I'd recommend revisiting this code to something like
```
if (!authHeader || !authHeader.startsWith("Bearer ")) {
   return res.status(403).send("Unauthorised");
}

const token = authHeader.substring(7, authHeader.length);
```

JamesMahy commented on src/middleware/auth-middleware.ts:
```diff
@@ -0,0 +1,28 @@
+
+const verifyToken = (req: Request, res: Response, next: NextFunction) => {
+    const authHeader = req.headers.authorization
+    let token;
```
should be const as we only set it once (see below)

JamesMahy commented on src/routes/auth-route.ts:
```diff
@@ -0,0 +1,11 @@
+
+const authController = new AuthController();
+authRouter.post('/registerUser', authController.registerUser)
+authRouter.post('/login', authController.login)
```
Think the router > controller > service is unncessarily verbose - I would put the controller code in here.

JamesMahy commented on src/routes/user-route.ts:
```diff
@@ -0,0 +1,12 @@
+
+const express = require('express');
+const userRouter = express.Router();
+//const userController = require('../controllers/user-controller');
```
Remove comment

JamesMahy commented on src/routes/auth-route.ts:
```diff
@@ -0,0 +1,11 @@
+const express = require('express');
```
Remove commonJs

JamesMahy commented on src/routes/user-route.ts:
```diff
@@ -0,0 +1,12 @@
+import {verifyToken} from "../middleware/auth-middleware";
+
+const express = require('express');
```
Remove commonjs

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+import {CreateUserRequest, LoginRequest} from "../global";
```
No need to import global

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+
+const prisma = new PrismaClient()
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
```
remove commonjs

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+const jwt = require('jsonwebtoken');
+const saltRounds = 10;
+
+export class AuthService {
```
This doesn't need to be a class, it's not retaining any state

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+const saltRounds = 10;
+
+export class AuthService {
+    registerUser = async (createUserRequest: CreateUserRequest) => {
```
Because this is a service, `firstName` and `email` should be passed in as independent parameters, this couples the service too tightly to the request and means this method is not easily reusable.

If the intention that it wouldn't be reusable, making it as a service just adds more unnecessary abstraction

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+
+export class AuthService {
+    registerUser = async (createUserRequest: CreateUserRequest) => {
+        const hashedPassword = await bcrypt.hash(createUserRequest.password, saltRounds);
```
If you don't change them to parameters, we should be destructing `createUserRequest`
EG:
`const { firstName, email, password } = createUserRequest;`

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+        })
+    }
+
+    login = async (loginRequest: LoginRequest) => {
```
Same as above, with passing the Request to the service

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+            }
+        })
+
+        if(!user) {
```
can be `if (!user) return;`

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+            return;
+        }
+
+        const validate = await bcrypt.compare(loginRequest.password, user.password)
```
Fairly sure `bcrypt.compare` will throw on error, you need to have error handling here

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+            userId: user.id,
+            email: user.email
+        }, process.env.TOKEN_KEY, {
+            expiresIn: "3h"
```
I would set this as a env value or as a constant at the top of the file

Awesomeluke1994 committed:
Fixed middleware to actually attempt to grab the payload if the decoding is a success.

If not is is picked up by the try and catch and throws a 401.

And holding the user details as request params.

JamesMahy commented on src/services/auth-service.ts:
```diff
@@ -0,0 +1,48 @@
+const prisma = new PrismaClient()
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const saltRounds = 10;
```
prefer these types of constants as ALL CAPS to make it clear where it's coming from

Awesomeluke1994 committed:
Changing the param value.

JamesMahy reviewed (commented):
Overall a good start, I would add eslint with `eslint:recommended` rules and prettier just to add some consistency to patterns

Awesomeluke1994 committed:
Allowed a way to pass through token details for context based usage.

Awesomeluke1994 committed:
Adjusted auth service - removed class definition and export as a functional service

Awesomeluke1994 committed:
Fixing my package.json - moving types to devDep

Awesomeluke1994 committed:
Added some validation for auth controller

Checking body has a email and body
Validating the email
Checking if a current user is registered using the email
Throwing a 500 if any error occurs when registering a user.

Awesomeluke1994 committed:
Tidied up auth route - getting it to export default instead

Set tokenDetails as optional

Removed some commonjs imports

Awesomeluke1994 committed:
Made it so password/firstname/lastname are not optional

Added the validation check to make sure that they exist in the body.

Tided up some commonjs imports

Awesomeluke1994 committed:
Making sure the validation of each body item works.

Awesomeluke1994 committed:
Removed last commonjs import

Awesomeluke1994 committed:
Added eslint

Awesomeluke1994 committed:
Tidied up the login service - added a try and catch around the compare method.

Awesomeluke1994 committed:
More tidying up of const values + exporting

Awesomeluke1994 committed:
Moved everything back to a global ts file

Actually veryfying the token :)

Awesomeluke1994 committed:
Added a try and catch around authService.login in auth controller in case login fails.

Awesomeluke1994 committed:
Adjusted the return of the getAllUsers to not return secret information

Awesomeluke1994 committed:
adding tidy up semi colons
